src/serac/physics/solid_residual.hpp

```diff
@@ -0,0 +1,629 @@
+        &shape_disp_space, &test_space, trial_spaces);
+  }
+
+  /**
+   * @brief register a custom boundary integral calculation as part of the residual
+   *
+   * @tparam active_parameters a list of indices, describing which parameters to pass to the q-function
+   * @param qfunction a callable that returns the traction on a boundary surface
+   * @param optional_domain The domain over which the boundary integral is evaluated. If nothing is supplied the entire
+   * boundary is used.
+   * ~~~ {.cpp}
+   *
+   *  solid_mechanics.addCustomBoundaryIntegral(DependsOn<>{}, [](double t, auto position, auto displacement, auto
+   * acceleration, auto shape){ auto [X, dX_dxi] = position;
+   *
+   *     auto [u, du_dxi] = displacement;
+   *     auto f           = u * 3.0 (X[0] < 0.01);
+   *     return f;  // define a displacement-proportional traction at a given support
+   *  });
+   *
+   * ~~~
+   *
+   * @note This method must be called prior to completeSetup()
+   */
+  template <int... active_parameters, typename callable>
+  void addCustomBoundaryIntegral(DependsOn<active_parameters...>, callable qfunction,
+                                 const std::optional<Domain>& optional_domain = std::nullopt)
+  {
+    Domain domain = (optional_domain) ? *optional_domain : EntireBoundary(mesh_);
+
+    residual_->AddBoundaryIntegral(Dimension<dim - 1>{}, DependsOn<0, 1, 2, active_parameters + NUM_FIELD_OFFSET...>{},
+                                   qfunction, domain);
+  }
+
+  /**
+   * @brief Set the material stress response and mass properties for the physics module
+   *
+   * @tparam MaterialType The solid material type
+   * @tparam StateType the type that contains the internal variables for MaterialType
+   * @param material A material that provides a function to evaluate stress
+   * @param domain Domain for material to be integrated over
+   * @pre material must be a object that can be called with the following arguments:
+   *    1. `MaterialType::State & state` an mutable reference to the internal variables for this quadrature point
+   *    2. `tensor<T,dim,dim> du_dx` the displacement gradient at this quadrature point
+   *    3. `tuple{value, derivative}`, a tuple of values and derivatives for each parameter field
+   *            specified in the `DependsOn<...>` argument.
+   *
+   * @note The actual types of these arguments passed will be `double`, `tensor<double, ... >` or tuples thereof
+   *    when doing direct evaluation. When differentiating with respect to one of the inputs, its stored
+   *    values will change to `dual` numbers rather than `double`. (e.g. `tensor<double,3>` becomes `tensor<dual<...>,
+   * 3>`)
+   *
+   * @param qdata the buffer of material internal variables at each quadrature point
+   *
+   * @pre MaterialType must have a public member variable `density`
+   * @pre MaterialType must define operator() that returns the Cauchy stress
```

Comment:
There are a few errors in the comment:

- The material must have a *method* to return the density (not a public member)
- The stress operator is called `pkStress`, not `operator()`
- The material returns PK, not Cauchy